gcc/d/expr.cc

```diff
@@ -2830,75 +2830,153 @@ build_expr(Expression *e, bool const_p)
+	  tree t = build_expr(vd->edtor);
+	  dtors = maybe_compound_expr(t, dtors);
 	}
     }
 
-  if (tdtors != NULL_TREE)
+  return dtors;
+}
+
+// Separate the value of a compound expression from the left hand side
+// expressions in RESULT.  Returns the result in EXPR and VALUE.
+
+static void
+get_compound_value(tree result, tree *expr, tree *value)
+{
+  if (TREE_CODE (result) == COMPOUND_EXPR)
+    {
+      *expr = TREE_OPERAND (result, 0);
+      *value = (TREE_OPERAND (result, 1));
+    }
+  else if (TREE_CODE (result) == INDIRECT_REF)
+    {
+      // Rewrite: *(e1, e2) => (e1, *e2)
+      tree ref = NULL_TREE;
+      get_compound_value(TREE_OPERAND (result, 0), expr, &ref);
+      *value = build_deref(ref);
+    }
+  else if (TREE_CODE (result) == ADDR_EXPR)
+    {
+      // Rewrite: &(e1, e2) => (e1, &e2)
+      tree val = NULL_TREE;
+      get_compound_value(TREE_OPERAND (result, 0), expr, &val);
+      *value = build_address(val);
+    }
+  else if (TREE_CODE (result) == NOP_EXPR)
     {
-      TOK rtoken = (e->op != TOKcomma) ? e->op : ((CommaExp *) e)->e2->op;
+      // Rewrite: cast(T)(e1, e2) => (e1, cast(T)e2)
+      tree val = NULL_TREE;
+      get_compound_value(TREE_OPERAND (result, 0), expr, &val);
+      *value = build_nop(TREE_TYPE (result), val);
+    }
+  else
+    *value = result;
+}
 
-      // For construction of temporaries, if the constructor throws, then
-      // we don't want to run the destructor on incomplete object.
-      CallExp *ce = (e->op == TOKcall) ? ((CallExp *) e) : NULL;
-      bool catchCtor = true;
-      if (ce != NULL && ce->e1->op == TOKdotvar)
-	{
-	  DotVarExp *dve = (DotVarExp *) ce->e1;
-	  if (dve->e1->op == TOKcomma && dve->var->isCtorDeclaration()
-	      && ((CommaExp *) dve->e1)->e1->op == TOKdeclaration
-	      && ((CommaExp *) dve->e1)->e2->op == TOKvar)
-	    catchCtor = false;
-	}
+// Same as build_expr, but also calls destructors on any temporaries.
+
+tree
+build_expr_dtor(Expression *e)
+{
+  // Codegen can be improved by determining if no exceptions can be thrown
+  // between the ctor and dtor, and eliminating the ctor and dtor.
+  size_t starti = cfun->language->vars_in_scope.length();
+  tree t = build_expr(e);
+  tree dtors = build_dtor_list(starti, cfun->language->vars_in_scope.length());
+
+  if (dtors != NULL_TREE)
+    {
+      // Split comma expressions, so that only the value is maybe saved.
+      tree expr = NULL_TREE;
+      tree value = NULL_TREE;
+      get_compound_value(t, &expr, &value);
 
-      // Wrap function/ctor and dtors in a try/finally expression.
-      if (catchCtor && (rtoken == TOKcall || rtoken == TOKnew))
+      // For construction of temporaries, if the constructor throws, then we
+      // don't want to run the destructor on the incomplete object.
+      CallExp *ce = (e->op == TOKcall) ? ((CallExp *) e) : NULL;
+      if (ce != NULL && ce->e1->op == TOKdotvar
+	  && ((DotVarExp *) ce->e1)->var->isCtorDeclaration())
```

Comment:
I guess it is here that triggered the change.  Before, only `TOKcall` and `TOKnew` expressions were wrapped in a try/finally. Now it is all expressions except constructors. 

I considered this reasonable, but maybe an exception could be made for temporaries as well.  I'll have to think on it.